module/builder/consensus/builder_test.go

```diff
@@ -730,3 +730,86 @@ func (bs *BuilderSuite) TestPayloadReceiptLimit() {
 	bs.Require().NoError(err)
 	bs.Assert().Equal(bs.pendingReceipts[:limit], bs.assembled.Receipts, "should have excluded receipts above maxReceiptCount")
 }
+
+// TestPayloadReceiptNoParentResult tests that the builder does not include
+// receipts whose PreviousResult is not already incorporated in the chain.
+//
+// Here we create 4 consecutive blocks S, A, B, and C, where A contains a valid
+// receipt for block S, but blocks B and C have empty payloads.
+//
+// We populate the mempool with valid receipts for blocks A, and C, but NOT for
+// block B.
+//
+// The expected behaviour, is that the builder should not include the receipt
+// for block C, because the chain and the mempool do not contain a valid receipt
+// for the parent result (block B's result).
+//
+// ... <- S <- A[ER{S}] <- B <- C <- X (candidate)
+func (bs *BuilderSuite) TestPayloadReceiptNoParentResult() {
+	// create and save block S
+	s := bs.blocks[bs.parentID]
+	sReceipt := unittest.ExecutionReceiptFixture(
+		unittest.WithResult(
+			unittest.ExecutionResultFixture(
+				unittest.WithBlock(s),
+			),
+		),
+	)
+
+	// create and save block A with a valid receipt for S
+	a := unittest.BlockWithParentFixture(s.Header)
+	a.SetPayload(flow.Payload{
+		Receipts: []*flow.ExecutionReceipt{sReceipt},
+	})
+	bs.headers[a.ID()] = a.Header
+	bs.blocks[a.ID()] = &a
+	bs.index[a.ID()] = a.Payload.Index()
```

Comment:
maybe we could extract a function so that this can be simplified into `bs.storageHasBlock(a)`